src/repr/src/adt/range.rs

```diff
@@ -27,136 +35,301 @@ bitflags! {
     }
 }
 
-/// A continuous set of domain values.
-#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, Ord, PartialOrd)]
-pub struct Range<'a> {
-    /// None value represents empty range
-    pub inner: Option<RangeInner<'a>>,
+/// A range of values along the domain `D`.
+///
+/// `D` is generic to facilitate interoperating over multiple representation,
+/// e.g. `Datum` and `mz_pgrepr::Value`. Because of the latter, we have to
+/// "manually derive" traits over `Range`.
+///
+/// Also notable, is that `Datum`s themselves store ranges as
+/// `Range<DatumNested<'a>>`, which lets us avoid unnecessary boxing of the
+/// range's finite bounds, which are most often expressed as `Datum`.
+pub struct Range<D> {
+    pub inner: Option<RangeInner<D>>,
```

Comment:
Prefer the option because it supports `map` w/o writing more code; will likely convert the Options to enums in another PR, though. Will add the comment back in.